src/emulate.c

```diff
@@ -356,107 +359,43 @@ RVOP(jalr, {
     return true;
 })
 
-/* BEQ: Branch if Equal */
-RVOP(beq, {
-    const uint32_t pc = rv->PC;
-    if (rv->X[ir->rs1] != rv->X[ir->rs2]) {
-        if (!ir->branch_untaken)
-            goto nextop;
-        rv->PC += ir->insn_len;
-        return ir->branch_untaken->impl(rv, ir->branch_untaken);
-    }
-    rv->PC += ir->imm;
-    /* check instruction misaligned */
-    RV_EXC_MISALIGN_HANDLER(pc, insn, false, 0);
-    if (ir->branch_taken)
-        return ir->branch_taken->impl(rv, ir->branch_taken);
+#define BRNACH_FUNC(type, comp)                                  \
+    const uint32_t pc = rv->PC;                                  \
+    if ((type) rv->X[ir->rs1] comp(type) rv->X[ir->rs2]) {       \
+        branch_take = false;                                     \
+        if (!ir->branch_untaken)                                 \
+            goto nextop;                                         \
+        rv->PC += ir->insn_len;                                  \
+        last_pc = rv->PC;                                        \
+        return ir->branch_untaken->impl(rv, ir->branch_untaken); \
+    }                                                            \
+    branch_take = true;                                          \
+    rv->PC += ir->imm;                                           \
+    /* check instruction misaligned */                           \
+    RV_EXC_MISALIGN_HANDLER(pc, insn, false, 0);                 \
```

Comment:
Is it possible to omit misalign handler for `BRANCH_FUNC`?

Reply:
No, we just define macro `BRANCH_FUNC` to avoid duplicate code, the implementation deos not change.